Sources/SwiftDocC/Infrastructure/DocumentationCurator.swift

```diff
@@ -242,19 +257,48 @@ struct DocumentationCurator {
+                                explanation: "\(topicSectionBaseExplanation). Modules should be roots in the documentation hierarchy."),
+                            possibleSolutions: removeListItemSolutions))
                         continue
                     }
                 }
                 
                 // Verify we are not creating a graph cyclic relationship.
                 guard childReference != nodeReference else {
-                    problems.append(Problem(diagnostic: Diagnostic(source: source(), severity: .warning, range: range(), identifier: "org.swift.docc.CyclicReference", summary: "A symbol can't link to itself from within its Topics group in \(nodeReference.absoluteString.singleQuoted)"), possibleSolutions: []))
+                    problems.append(Problem(
+                        diagnostic: Diagnostic(
+                            source: source(), severity: .warning, range: range(), identifier: "org.swift.docc.CyclicReference",
+                            summary: "Organizing \(describeForDiagnostic(childReference).singleQuoted) under itself forms a cyclic documentation hierarchy",
+                            explanation: "\(topicSectionBaseExplanation). The documentation hierarchy shouldn't contain cycles."),
+                        possibleSolutions: removeListItemSolutions
+                    ))
                     continue
                 }
                 
                 guard !isReference(childReference, anAncestorOf: nodeReference) else {
-                    problems.append(Problem(diagnostic: Diagnostic(source: source(), severity: .warning, range: range(), identifier: "org.swift.docc.CyclicReference", summary: "Linking to \((link.destination ?? "").singleQuoted) from a Topics group in \(nodeReference.absoluteString.singleQuoted) isn't allowed", explanation: "The former is an ancestor of the latter"), possibleSolutions: []))
+                    // Adding this edge in the topic graph _would_ introduce a cycle.
+                    // In order to produce more actionable diagnostics, create a new graph that has this cycle.
+                    var edges = context.topicGraph.edges
+                    edges[nodeReference, default: []].append(childReference)
+                    let graph = DirectedGraph(edges: edges)
+                    
+                    let cycleDescriptions: [String] = graph.cycles(from: nodeReference).map { prettyPrint(cycle: $0) }
+                    
+                    problems.append(Problem(
+                        diagnostic: Diagnostic(
+                            source: source(), severity: .warning, range: range(), identifier: "org.swift.docc.CyclicReference",
+                            summary: "Organizing \(describeForDiagnostic(childReference).singleQuoted) under \(describeForDiagnostic(nodeReference).singleQuoted) forms a cyclic documentation hierarchy",
```

Comment:
Like above, maybe simplify to "...forms N cycles." - or "forms a cycle." ... checking `cycleDescriptions` like below.